docs/plugin-protocol/tfplugin5.3.proto

```diff
@@ -0,0 +1,381 @@
+        bool deprecated = 6;
+    }
+
+    message Attribute {
+        string name = 1;
+        bytes type = 2;
+        string description = 3;
+        bool required = 4;
+        bool optional = 5;
+        bool computed = 6;
+        bool sensitive = 7;
+        StringKind description_kind = 8;
+        bool deprecated = 9;
+    }
+
+    message NestedBlock {
+        enum NestingMode {
+            INVALID = 0;
+            SINGLE = 1;
+            LIST = 2;
+            SET = 3;
+            MAP = 4;
+            GROUP = 5;
+        }
+
+        string type_name = 1;
+        Block block = 2;
+        NestingMode nesting = 3;
+        int64 min_items = 4;
+        int64 max_items = 5;
+    }
+
+    // The version of the schema.
+    // Schemas are versioned, so that providers can upgrade a saved resource
+    // state when the schema is changed.
+    int64 version = 1;
+
+    // Block is the top level configuration block for this schema.
+    Block block = 2;
+}
+
+service Provider {
+    //////// Information about what a provider supports/expects
+    rpc GetSchema(GetProviderSchema.Request) returns (GetProviderSchema.Response);
+    rpc PrepareProviderConfig(PrepareProviderConfig.Request) returns (PrepareProviderConfig.Response);
+    rpc ValidateResourceTypeConfig(ValidateResourceTypeConfig.Request) returns (ValidateResourceTypeConfig.Response);
+    rpc ValidateDataSourceConfig(ValidateDataSourceConfig.Request) returns (ValidateDataSourceConfig.Response);
+    rpc UpgradeResourceState(UpgradeResourceState.Request) returns (UpgradeResourceState.Response);
+
+    //////// One-time initialization, called before other functions below
+    rpc Configure(Configure.Request) returns (Configure.Response);
+
+    //////// Managed Resource Lifecycle
+    rpc ReadResource(ReadResource.Request) returns (ReadResource.Response);
+    rpc PlanResourceChange(PlanResourceChange.Request) returns (PlanResourceChange.Response);
+    rpc ApplyResourceChange(ApplyResourceChange.Request) returns (ApplyResourceChange.Response);
+    rpc ImportResourceState(ImportResourceState.Request) returns (ImportResourceState.Response);
+
+    rpc ReadDataSource(ReadDataSource.Request) returns (ReadDataSource.Response);
+
+    //////// Graceful Shutdown
+    rpc Stop(Stop.Request) returns (Stop.Response);
+}
+
+message GetProviderSchema {
+    message Request {
+    }
+    message Response {
+        Schema provider = 1;
+        map<string, Schema> resource_schemas = 2;
+        map<string, Schema> data_source_schemas = 3;
+        repeated Diagnostic diagnostics = 4;
+        Schema provider_meta = 5;
+        Capabilities capabilities = 6;
+    }
+
+
+    // Capabilities allows providers to communicate extra information regarding
+    // supported protocol features. This is used to indicate availability of
+    // certain forward-compatible changes which may be optional in a major
+    // protocol version, but cannot be tested for directly.
+    message Capabilities {
```

Comment:
The reason we added this was specifically because of terraform-plugin-go (and to a lesser extent, theoretical providers built directly on grpc). While both the legacy shims and the framework handle this case correctly, we couldn't rule out unknown providers which don't. Since capabilities are essentially simple flags, I'm not sure what other design you had in mind. The choice for boolean fields was somewhat arbitrary, but IMO does make the resulting client code easier to read than a series of enums which need to be iterated over individually, though I have no strong opposition if there's another reason to change the data structure.

The basis for using the overall style of protobuf+grpc was specifically for backwards compatible additions whenever possible. While the alternative here was a new method, `PlanResourceDestroy`, that seemed unnecessary for the small edge case of an unsupported provider, and broke from the consistency of the overall API.

Reply:
I think our experience with protocol version 6 led us to conclude that we basically never want to have another major version of the protocol again, and that backward-compatible additions to protocol version 6 are going to be our main focus moving forward.

Granular capability negotiation like this feels like a well-trodden path with other similar protocols, but I will admit that I didn't interrogate that deeply and just suggested it because it was familiar to me as a solution to similar problems elsewhere.

While I can understand that it might feel quite drastic to introduce something entirely new here, I think in practice the cost of us being "wrong" about this is relatively low: if we never use this capability mechanism ever again then it will be slightly annoying to have this extra message type here in the protocol specification forever, but I don't think it will have any cross-cutting impact to any other codepaths and therefore we can safely ignore it if we find that it's never useful again in practice, or if we find a better way in future.

Reply:
Putting these things together, a breaking change to the protocol now involves incrementing the minor version and adding a new Capabilities field. These can never be removed or changed. I agree with Martin that this is at worst annoying and that the general granular client/server capabilities design is well known.